runtime/major_gc.c

```diff
@@ -116,34 +129,146 @@ static unsigned long major_gc_counter = 0;
+    caml_skiplist_insert(&chunk_sklist, (uintnat)heap_chunk,
+                          (uintnat)(heap_chunk+Chunk_size(heap_chunk)));
+    heap_chunk = Chunk_next(heap_chunk);
+  } while( heap_chunk != NULL );
+
+  for( entry = 0; entry < mark_stack_count ; entry++ ) {
+    mark_entry me = mark_stack[entry];
+    value* block_op = Op_val(me.block);
+    uintnat chunk_addr = 0, chunk_addr_below = 0;
+
+    if( caml_skiplist_find_below(&chunk_sklist, (uintnat)me.block,
+          &chunk_addr, &chunk_addr_below)
+        && me.block < chunk_addr_below ) {
+
+      if( Chunk_redarken_start(chunk_addr) > block_op ) {
+        Chunk_redarken_start(chunk_addr) = block_op;
+      }
+
+      if( Chunk_redarken_end(chunk_addr) < block_op ) {
+        Chunk_redarken_end(chunk_addr) = block_op;
+      }
+
+      if( redarken_first_chunk == NULL
+          || redarken_first_chunk > (char*)chunk_addr ) {
+        redarken_first_chunk = (char*)chunk_addr;
+      }
+    }
+  }
+
+  caml_skiplist_empty(&chunk_sklist);
 
-  CAMLassert (gray_vals_cur == gray_vals_end);
-  if (gray_vals_size < Caml_state->stat_heap_wsz / 32){
-    caml_gc_message (0x08, "Growing gray_vals to %"
+  caml_gc_message(0x08, "Mark stack overflow.\n");
+
+  stk->count = 0;
+}
+
+static void realloc_mark_stack (struct mark_stack* stk)
+{
+  mark_entry* new;
+  uintnat mark_stack_bsize = stk->size * sizeof(mark_entry);
+
+  if ( Wsize_bsize(mark_stack_bsize) < Caml_state->stat_heap_wsz / 32 ) {
+    caml_gc_message (0x08, "Growing mark stack to %"
                            ARCH_INTNAT_PRINTF_FORMAT "uk bytes\n",
-                     (intnat) gray_vals_size * sizeof (value) / 512);
-    new = (value *) caml_stat_resize_noexc ((char *) gray_vals,
-                                            2 * gray_vals_size *
-                                            sizeof (value));
-    if (new == NULL){
-      caml_gc_message (0x08, "No room for growing gray_vals\n");
-      gray_vals_cur = gray_vals;
-      heap_is_pure = 0;
-    }else{
-      gray_vals = new;
-      gray_vals_cur = gray_vals + gray_vals_size;
-      gray_vals_size *= 2;
-      gray_vals_end = gray_vals + gray_vals_size;
+                     (intnat) mark_stack_bsize * 2 / 1024);
+
+    new = (mark_entry*) caml_stat_resize_noexc ((char*) stk->stack,
+                                                2 * mark_stack_bsize);
+    if (new != NULL) {
+      stk->stack = new;
+      stk->size *= 2;
+      return;
     }
-  }else{
-    gray_vals_cur = gray_vals + gray_vals_size / 2;
-    heap_is_pure = 0;
   }
+
+  caml_gc_message (0x08, "No room for growing mark stack. Pruning..\n");
+  mark_stack_prune(stk);
 }
 
+/* This function pushes the provided mark_entry [me] onto the current mark
+   stack [stk]. It first checks, if the block is small enough, whether there
+   are any fields we would actually do mark work on. If so then it enqueues
+   the entry. */
+Caml_inline void mark_stack_push(struct mark_stack* stk, mark_entry me,
```

Comment:
Maybe it's just a superstition of mine, but I dislike passing structs by value.  For one thing, it's hard to know which more or less efficient calling conventions is dictated by the ABI.  I would have passed the two components of `me` as two scalar arguments.

Reply:
Fixed in most recent commit.